### DISCUSSION TITLE
Alternative firmware update of a STM32 using USB mass storage device class

### DISCUSSION THREAD

**Discussion by saraverbeecke** [micropython/micropython#13731] [STM32 / Pyboard] [answered]
It is possible to add files onto the filesystem when plugging the Pyboard in the PC using USB.
I would like to make it possible for end-users to save a firmware .bin / .hex file to the mass storage device in order to update the firmware without using either DFU mode, stlink or OpenOCD.
Is this possible?

**Comment by dhalbert** [comment]
You could install the TinyUF2 bootloader and upload UF2 files. It has support for several STM32 variants. https://github.com/adafruit/tinyuf2. The MicroPython build for the PyBoard might have to be adapted in some way: start address, reserve space for the bootloader, etc. TinyUF2 is not protected like, say, a SAMD bootloader. It is a secondary bootloader.

This has been done for MicroPython on ESP32-xx: https://github.com/micropython/micropython/issues/8222

**Comment by jimmo** [comment, marked as answer]
@saraverbeecke What Dan said above is probably what you want.

For completeness though, the MicroPython project also provides a custom bootloader for stm32 that we need for the pyboard-d series, but you can also install it on most stm32 boards (including the pyboard v1.x).

It supports installing firmware from a file on the filesystem. So the workflow is:
1. Copy the firmware to the device filesystem (e.g. via USB MSC, mpremote, your own python code)
2. Trigger a reboot into mboot which will install the firmware for you.

I've seen people use this for OTA updates. It supports compressed firmware images and image signing. See https://github.com/micropython/micropython/tree/master/ports/stm32/mboot (and in particular `fwupdate.py` which shows how to trigger mboot).

Like Dan said, it requires the main firmware to be mboot-aware (because offsets and start addresses change etc). The PYBV11 board definition has an example of how to do this: https://github.com/micropython/micropython/blob/master/ports/stm32/boards/PYBV11/mpconfigboard.mk

**Reply by saraverbeecke** [reply]
Thanks @jimmo I was not aware of this info on the Micropython Github. The official docs don't mention mboot options, maybe an idea to add these there as well?

**Reply by saraverbeecke** [reply]
@jimmo i was looking into this again, and i am confused. Do i really need mboot to update the firmware using a new firmware file from the internal file system? Looking at the `fwupdate.py` it seems that i can just check during boot in e.g. boot.py if there is any firmware file on the fs available (can be put there by the end-user using usb (MSC)) and then update the firmware using that file and dealing with it as done in the `fwupdate.py` file, or do i miss something here?

**Reply by andrewleech** [reply]
fwupdate does have functions to write directly to flash which can be used to update mboot if needed. 

If you instead try to directly update the main firmware this way however, as it's writing it will eventually replace the actual running micropython code / vm and inevitably crash. 

mboot is a separate codebase that can safely run and update the main firmware without overwriting itself.

**Reply by saraverbeecke** [reply]
Ah ok, thank you Andrew for the clarification. Say I have Mboot enabled and I booted in the bootloader, how can I make it start doing an update of the application firmware from a new firmware file placed on the internal file system?

**Reply by andrewleech** [reply]
Typically you trigger the install process from your application code, I think fwupdate might have the function for this already? 

Basically it stores the desired filename and details about the filesystem it's on in non -volatile ram then triggers the reboot into mboot. 
mboot then sees the trigger data and starts the update automatically.

**Reply by saraverbeecke** [reply]
Thank you for your answer Andrew, that is already some useful info. I am still struggling though, but got my expanded file system space working already. You can check my progress and extra questions in another opened discussion: [https://github.com/orgs/micropython/discussions/15193](https://github.com/orgs/micropython/discussions/15193).
